### PR TITLE
don't initialize the output embeddings if we're going to tie them to input embeddings

### PR DESCRIPTION
**Pull request by tom-p-reichel**
# What does this PR do?

This small change marks the output embeddings for a model as initialized if we will be tying them to the input embeddings. Without this change, the output embeddings are usually randomly initialized every time affected models (models that tie the output embeddings to input embeddings and do not otherwise initialize the output embeddings) are loaded. This seems to be responsible for *multiple second* startup delays in downstream tools, e.g. insanely-fast-whisper, as every single time the whisper model is loaded a very massive matrix is unnecessarily filled with uniformly random numbers before it is replaced with another matrix.

Before and after applying this patch, downstream tool insanely-fast-whisper transcribed a short audio file in 18 and 13 seconds respectively for a 5 second improvement. The patch does not seem to change the behavior of the tool-- a test transcription of an hour of audio remains unchanged before and after the patch.

I suspect other applications using models that tie their input/output embeddings together will experience a small speedup in loading from this patch.

I ran a portion of the transformers testing locally, which passed, but we'll see how the full test suite fares soon enough.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.

### CODE REVIEW

**Comment by amyeroberts** [comment]
Hi @tom-p-reichel - thanks for opening this PR! 

Very nice if there's such a big speedup 🔥  cc @sanchit-gandhi 

Could you share some snippets on how this was tested - speed and outputs?

**Comment by tom-p-reichel** [comment]
@amyeroberts Sure. Here's a minimal example:
`python -m timeit -s "from transformers import pipeline; import torch" "pipeline('automatic-speech-recognition',model='openai/whisper-large-v3', torch_dtype=torch.float16)"`

Before patch (origin/main):
`1 loop, best of 5: 6.91 sec per loop`

After patch:
`1 loop, best of 5: 1.72 sec per loop`

Note: If we're not using `dtype=torch.float16` as insanely-fast-whisper does then both versions run much faster and the patch only gives a marginal improvement.  It seems like randomly initializing something in float16 is really expensive!

As a sanity check that nothing terribly bad has happened to the weights as a result of this patch, we can also download some public domain audio and test that the downstream tool works as intended:

```
wget https://www.archive.org/download/metamorphosis_librivox/metamorphosis_librivox_64kb_mp3.zip
unzip metamorphosis_librivox_64kb_mp3.zip
insanely-fast-whisper --file-name metamorphosis_1_kafka_64kb.mp3
```
The resultant `output.json` pre and post patch are identical for the 49 minutes of audio transcribed.

**Comment by tom-p-reichel** [comment]
Rebase took care of the one failing documentation test. Anything else that needs to be done for this PR?

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for adding this and including all the details of how it was tested ❤️ 

Just want a final second check from @ArthurZucker who knows more about weight init logic

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -4001,6 +4001,11 @@ def _fix_key(key):
                 else:
                     _loaded_keys = loaded_keys
                 set_initialized_submodules(model, _loaded_keys)
+                # if we're about to tie the output embeds to the input embeds we don't need to init them
+                if hasattr(model.config, "tie_word_embeddings") and model.config.tie_word_embeddings:
+                    output_embeddings = model.get_output_embeddings()
+                    if output_embeddings is not None:
+                        output_embeddings._is_hf_initialized = True
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:4008]: Very nice!
I think we can also add something like https://github.com/huggingface/transformers/blob/90224dd59e92e11d99b5b09be84d3fe7794636b9/src/transformers/modeling_utils.py#L1580

for 
```python
        if getattr(self.config, "is_encoder_decoder", False) and getattr(self.config, "tie_encoder_decoder", False):
            if hasattr(self, self.base_model_prefix):
                self = getattr(self, self.base_model_prefix)
            self._tie_encoder_decoder_weights(self.encoder, self.decoder, self.base_model_prefix)
```
which can be tested on MarianMT models but should come in another PR!

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the PR! 
I think we can add a very small test to make sure fast init doesn't mess up anything near https://github.com/huggingface/transformers/blob/90224dd59e92e11d99b5b09be84d3fe7794636b9/tests/test_modeling_common.py#L434 

Otherwise LGTM! 🤗

**Comment by tom-p-reichel** [comment]
Hi, sorry for the delay, I was doing graduate student things. Here is a rebase onto the current main and a new test.

@ArthurZucker is this what you intended for the test case? There are already test cases that ensure that the model is unchanged while loading and saving, even during fast_init, so I only test the new behavior I added -- which is that the output embedding is never initialized if it's tied.

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks for the test, exactly what I wanted! 🤗 

let's rebase one last time and CIs should go green

**Comment by tom-p-reichel** [comment]
@ArthurZucker Good to hear it, rebased.

**Comment by ArthurZucker** [comment]
Thanks for the PR!

**Comment by ArthurZucker** [comment]
FYI @pacman100 if we have any deepspeed issues 🤗

**Comment by hackyon** [comment]
FYI I ran into an issue with this for the ElectraModel with this [test](https://github.com/huggingface/transformers/blob/d628664688b05cabdd69f4e7e295bc4aee0a8d31/tests/test_modeling_common.py#L434). For some reason, the failure only showed up when I modified the source for ElectraModel for an unrelated change. 

The issue appears to be that the `bias` term in the `output_embeddings` is not tied in [`_tie_or_clone_weights()`](https://github.com/huggingface/transformers/blob/d628664688b05cabdd69f4e7e295bc4aee0a8d31/src/transformers/modeling_utils.py#L1712). The test ended up failing because initialization was skipped for the `bias` as well.

I added this [commit](https://github.com/huggingface/transformers/pull/28802/commits/9fa20df3faba1a72fb21cb7eca55e4065836a5f3) to get the test to pass. Let me know if you have any concerns. AFAIT, the whisper model doesn't have `bias` in their embeddings, and so my fix shouldn't affect the gains from this change.

**Comment by tom-p-reichel** [comment]
@hackyon Good catch! Yes, that is an oversight. Curious about needing to edit ElectraModel to see the failure and the passing tests in this PR.

**Pull request merged**